content/oci/oci.go

```diff
@@ -0,0 +1,227 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package oci
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/opencontainers/go-digest"
+	specs "github.com/opencontainers/image-spec/specs-go"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/internal/graph"
+	reslv "oras.land/oras-go/v2/internal/resolver"
+)
+
+// ociImageIndexFile is the file name of the index from the OCI Image Layout Specification.
+// Reference: https://github.com/opencontainers/image-spec/blob/master/image-layout.md#indexjson-file
+const ociImageIndexFile = "index.json"
+
+// Store implements `oras.Target`, and represents a content store
+// based on file system with the OCI-Image layout.
+// Reference: https://github.com/opencontainers/image-spec/blob/master/image-layout.md
+type Store struct {
+	storage  content.Storage
+	resolver *reslv.Memory
+	finder   *graph.Memory
+	index    *ocispec.Index
+
+	root      string
+	indexPath string
+}
+
+func New(root string) (*Store, error) {
+	store := &Store{
+		storage:   NewStorage(root),
+		resolver:  reslv.NewMemory(),
+		finder:    graph.NewMemory(),
+		root:      root,
+		indexPath: filepath.Join(root, ociImageIndexFile),
+	}
+
+	if err := store.validateOCILayoutFile(); err != nil {
+		return nil, err
+	}
+
+	if err := store.loadIndex(); err != nil {
+		return nil, err
+	}
```

Comment:
What component is supposed to do the initialization to create the layerout and index file?

Reply:
The layout file is initialized in [validateOCILayoutFile()](https://github.com/Wwwsylvia/oras-go/blob/oci_target/content/oci/oci.go#L155) when loading. The index file will be initialized on first tag in [saveIndex](https://github.com/Wwwsylvia/oras-go/blob/oci_target/content/oci/oci.go#L230)

Reply:
I see. It is not clear from the name that `validateOCILayeroutFile` creates the layout file if it doesn't exist. Maybe you can rename it something like validateOrCreateOCILayoutFile

Reply:
Do you know where I can find a code snippet example of copy an oci artifact from registry to local?

Reply:
We shall update the [examples](https://github.com/Wwwsylvia/oras-go/tree/oci_target/examples) once the implementation is done... Before that, @akashsinghal has written some code to copy some artifacts from OCI target to remote registry, which is helpful - https://github.com/akashsinghal/accelerated-container-image/blob/47ef7a728bed74020ef48814e8efb461f3db0236/cmd/ctr/overlaybd_conv.go#L740-L781

Reply:
Linking to #60